### ISSUE TITLE
metadata::filter_platform fails on !amd64

### ISSUE THREAD

**Issue by mwhudson** [rust-lang/cargo#9238]
**Problem**
metadata::filter_platform fails on e.g. powerpc64le-linux-gnu

**Steps**

1. install rustup
2. rustup default stable
3. git clone https://github.com/rust-lang/cargo
4. cd cargo
5. cargo test metadata::filter_platform

**Possible Solution(s)**

The failure is rather long (you can see it in https://launchpadlibrarian.net/526141668/buildlog_ubuntu-hirsute-amd64.cargo_0.51.0-0ubuntu1~ppa1_BUILDING.txt.gz) but the upshot is that /packages/2/dependencies/2 and /packages/2/dependencies/3 are the other way around from how the test expects (test expects alt-dep first, output has host-dep first).

Now, I'm no expert but I don't see anything in the code that guarantees the ordering of dependencies in the output. So I'm guessing this only passes on amd64 by chance. But I could be wrong!

**Notes**

Output of `cargo version`:

I observed this with c69409658a1f19452b784f675984e46da752a75b and 0.51.0 built with "rustc 1.50.0 (cb75ad5db 2021-02-10)" on Ubuntu 20.04. I also saw the same test building cargo 0.51.0 on arm64, s390x and armhf with my rustc 1.50.0 packages.

**Comment by ehuss** [comment]
I don't see any errors in the linked build log, but I was able to verify with an aarch64 system.

Yea, it looks like the `[target]` tables are stored in sorted order.  The test as written is assuming the host triple sorts alphabetically after `wasm32`, which is true for all `x86*` targets.

It'll be a bit tricky to fix, since there isn't any existing support code for ignoring the order of JSON arrays (I think it was removed in #8489).  I'm guessing adding an option to ignore array order is an possibility.  Another is to change that test to extract the JSON object, and selectively compare the important parts.
